kernel/kernel-api/src/main/java/io/delta/kernel/internal/snapshot/SnapshotManager.java

```diff
@@ -419,192 +401,180 @@ private LogSegment constructLogSegmentFromFileList(
+    // Step 8: Determine the version of the snapshot we can now load. //
+    ////////////////////////////////////////////////////////////////////
 
     final List<Long> deltaVersionsAfterCheckpoint =
         deltasAfterCheckpoint.stream()
             .map(fileStatus -> FileNames.deltaVersion(new Path(fileStatus.getPath())))
             .collect(Collectors.toList());
 
-    logDebug(
-        () -> format("deltaVersions: %s", Arrays.toString(deltaVersionsAfterCheckpoint.toArray())));
-
     final long newVersion =
         deltaVersionsAfterCheckpoint.isEmpty()
-            ? newCheckpointOpt.get().version
+            ? latestCompleteCheckpointVersion
             : ListUtils.getLast(deltaVersionsAfterCheckpoint);
 
-    // There should be a delta file present for the newVersion that we are loading
-    // (Even if `deltasAfterCheckpoint` is empty, `deltas` should not be)
-    if (deltas.isEmpty()
-        || FileNames.deltaVersion(deltas.get(deltas.size() - 1).getPath()) < newVersion) {
+    logger.info("New version to load: {}", newVersion);
+
+    /////////////////////////////////////////////
+    // Step 9: Perform some basic validations. //
+    /////////////////////////////////////////////
+
+    if (!latestCompleteCheckpointOpt.isPresent() && deltasAfterCheckpoint.isEmpty()) {
       throw new InvalidTableException(
-          tablePath.toString(), String.format("Missing delta file for version %s", newVersion));
+          tablePath.toString(), "No complete checkpoint found and no delta files found");
     }
 
-    versionToLoadOpt
-        .filter(v -> v != newVersion)
-        .ifPresent(
-            v -> {
-              throw DeltaErrors.versionAfterLatestCommit(tablePath.toString(), v, newVersion);
-            });
+    // If there's a checkpoint at version N then we also require that there's a delta file at that
+    // version.
+    if (latestCompleteCheckpointOpt.isPresent()
+        && listedDeltaFileStatuses.stream()
+            .map(x -> FileNames.deltaVersion(new Path(x.getPath())))
+            .noneMatch(v -> v == latestCompleteCheckpointVersion)) {
+      throw new InvalidTableException(
+          tablePath.toString(),
+          String.format("Missing delta file for version %s", latestCompleteCheckpointVersion));
+    }
 
-    // We may just be getting a checkpoint file after the filtering
-    if (!deltaVersionsAfterCheckpoint.isEmpty()) {
-      // If we have deltas after the checkpoint, the first file should be 1 greater than our
-      // last checkpoint version. If no checkpoint is present, this means the first delta file
-      // should be version 0.
-      if (deltaVersionsAfterCheckpoint.get(0) != newCheckpointVersion + 1) {
-        throw new InvalidTableException(
-            tablePath.toString(),
-            String.format(
-                "Unable to reconstruct table state: missing log file for version %s",
-                newCheckpointVersion + 1));
-      }
```

Comment:
This check is now done as part of verifyDeltaVersions I think?

Reply:
I think if we're throwing `InvalidTableExceptions` inside of `verifyDeltaVersions` the docs for that method need to be very specific to explain the input/output (what they are in the flow of the segment construction) and why they are invalid tables

Reply:
> This check is now done as part of verifyDeltaVersions I think?

Exactly!